biogeochem/EDPhysiologyMod.F90

```diff
@@ -1245,92 +1254,322 @@ subroutine phenology_leafonoff(currentSite)
+
+    dummy_n = 1.0_r8 ! make n=1 to get area of one tree.
+    spread = 1.0_r8  ! fix this to 0 to remove dynamics of canopy closure, assuming a closed canopy.
+    ! n.b. the value of this will only affect 'n', which isn't/shouldn't be a diagnostic in 
+    ! SP mode. 
+    call carea_allom(currentCohort%dbh,dummy_n,spread,currentCohort%pft,currentCohort%c_area)
+
+    !------------------------------------------
+    !  Calculate canopy N assuming patch area is full
+    !------------------------------------------
+    currentCohort%n = parea / currentCohort%c_area
+
+    ! correct c_area for the new nplant
+    call carea_allom(currentCohort%dbh,currentCohort%n,spread,currentCohort%pft,currentCohort%c_area)
+
+    ! ------------------------------------------
+    ! Calculate leaf carbon from target treelai
+    ! ------------------------------------------
+    currentCohort%treelai = tlai
+    canopylai(:) = 0._r8
+    leaf_c = leafc_from_treelai( currentCohort%treelai, currentCohort%pft, currentCohort%c_area,&
+         currentCohort%n, currentCohort%canopy_layer, currentCohort%vcmax25top)
+
+    !check that the inverse calculation of leafc from treelai is the same as the 
+    ! standard calculation of treelai from leafc. Maybe can delete eventually?
+
+    check_treelai = tree_lai(leaf_c, currentCohort%pft, currentCohort%c_area, &
+         currentCohort%n, currentCohort%canopy_layer,               &
+         canopylai,currentCohort%vcmax25top )
+
+    if( abs(currentCohort%treelai-check_treelai).gt.1.0e-12)then !this is not as precise as nearzero
+       write(fates_log(),*) 'error in validate treelai',currentCohort%treelai,check_treelai,currentCohort%treelai-check_treelai
+       call endrun(msg=errMsg(sourcefile, __LINE__))
+    end if
+
+    ! the carea_allom routine sometimes generates precision-tolerance level errors in the canopy area 
+    ! these mean that the canopy area does not exactly add up to the patch area, which causes chaos in 
+    ! the radiation routines.  Correct both the area and the 'n' to remove error, and don't use 
+    !! carea_allom in SP mode after this point. 
+
+    if(abs(currentCohort%c_area-parea).gt.nearzero)then ! there is an error
+       if(abs(currentCohort%c_area-parea).lt.10.e-9)then !correct this if it's a very small error
+          oldcarea = currentCohort%c_area
+          !generate new cohort area
+          currentCohort%c_area = currentCohort%c_area - (currentCohort%c_area- parea)
+          currentCohort%n = currentCohort%n * (currentCohort%c_area/oldcarea)
+          if(abs(currentCohort%c_area-parea).gt.nearzero)then
+             write(fates_log(),*) 'SPassign, c_area still broken',currentCohort%c_area-parea,currentCohort%c_area-oldcarea
+             call endrun(msg=errMsg(sourcefile, __LINE__))
+          end if
+       else 
+          write(fates_log(),*) 'SPassign, big error in c_area',currentCohort%c_area-parea,currentCohort%pft
```

Comment:
worried about logging that is not followed by an endrun.  Should we encapsulate this with a local debug flag that is default set to false?